flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2527,6 +2527,39 @@ class ScalarArrayExprLowering {
     return lexv;
   }
 
+  ExtValue userAssign(mlir::FuncOp userAssignment,
+                      const Fortran::lower::SomeExpr &lhs,
+                      const Fortran::lower::SomeExpr &rhs) {
+    semant = ConstituentSemantics::RefTransparent;
+    // 1) Lower the rhs expression with array_fetch op(s).
+    auto rexv = lower(rhs);
+    elementalValue = fir::getBase(rexv);
+    // 2) Lower the lhs expression to an array_update.
+    semant = ConstituentSemantics::CustomCopyInCopyOut;
+    auto lexv = lower(lhs);
+    bool isIllFormedLHS = false;
+    // 3) Insert the call
+    if (auto modifyOp = mlir::dyn_cast<fir::ArrayModifyOp>(
+            fir::getBase(lexv).getDefiningOp())) {
+      auto oldInnerArg = modifyOp.sequence();
+      auto offset = expSpace.argPosition(oldInnerArg);
+      expSpace.setInnerArg(offset, fir::getBase(lexv));
+      genScalarUserDefinedAssignmentCall(builder, getLoc(), userAssignment,
+                                         modifyOp.getResult(0), elementalValue);
+    } else {
+      // Ill-formed lhs. Still insert user assignment call for consistency.
+      isIllFormedLHS = true;
+      genScalarUserDefinedAssignmentCall(builder, getLoc(), userAssignment,
+                                         fir::getBase(lexv), elementalValue);
+    }
+    // 4) Finalize the inner context.
+    expSpace.finalizeContext();
+    // 5). Thread the array value updated forward.
+    if (!isIllFormedLHS)
+      builder.create<fir::ResultOp>(getLoc(), fir::getBase(lexv));
```

Comment:
Is skipping the terminator op a supported path?

Reply:
It is the case for the normal assignment (the terminator op is not generated for ill formed LHS): https://github.com/flang-compiler/f18-llvm-project/blob/ae9218ca2354c0cb62d77b7a529632f486102469/flang/lib/Lower/ConvertExpr.cpp#L2479.